File: compiler/ocaml/src/llvm_gen/runtime_link.ml
(* Runtime_link — ランタイムライブラリとのリンク支援 (Phase 1-5)
 *
 * このファイルは docs/plans/bootstrap-roadmap/1-5-runtime-integration.md §6.3 に基づき、
 * 生成した LLVM IR とランタイムライブラリをリンクする機能を提供する。
 *
 * 設計方針:
 * - ランタイムライブラリ (libreml_runtime.a) の検索
 * - プラットフォーム検出（macOS / Linux）に基づくリンカー設定
 * - Phase 1 スコープ: 静的リンクのみ対応
 *
 * 参考資料:
 * - docs/plans/bootstrap-roadmap/1-5-runtime-integration.md §6.3
 * - runtime/native/Makefile
 *)

(* ========== エラー型 ========== *)

exception LinkError of string

let link_error msg = raise (LinkError msg)
let link_errorf fmt = Printf.ksprintf link_error fmt

(* ========== プラットフォーム検出 ========== *)

type platform = MacOS | Linux | Windows | Unknown

let detect_platform () =
  let uname =
    try
      let ic = Unix.open_process_in "uname -s" in
      let result = input_line ic in
      let _ = Unix.close_process_in ic in
      result
    with _ -> "Unknown"
  in
  match uname with
  | "Darwin" -> MacOS
  | "Linux" -> Linux
  | "Windows_NT" | "CYGWIN" | "MINGW" -> Windows
  | _ -> Unknown

(* ========== ランタイムライブラリの検索 ========== *)

(** ランタイムライブラリのデフォルトパスを取得
 *
 * 優先順位:
 * 1. 環境変数 REML_RUNTIME_PATH が設定されている場合はそれを使用
 * 2. runtime/native/build/libreml_runtime.a（ローカルビルド）
 * 3. /usr/local/lib/reml/libreml_runtime.a（インストール版）
 *)
let find_runtime_library () =
  (* 1. 環境変数チェック *)
  let env_path =
    try Some (Sys.getenv "REML_RUNTIME_PATH") with Not_found -> None
  in
  match env_path with
  | Some path when Sys.file_exists path -> path
  | _ ->
      (* 2. ローカルビルドパス *)
      let local_path = "runtime/native/build/libreml_runtime.a" in
      if Sys.file_exists local_path then local_path
      else
        (* 3. インストール版パス *)
        let installed_path = "/usr/local/lib/reml/libreml_runtime.a" in
        if Sys.file_exists installed_path then installed_path
        else
          link_errorf
            "ランタイムライブラリが見つかりません。\n\
             以下のいずれかを試してください:\n\
             1. runtime/native で 'make runtime' を実行\n\
             2. 環境変数 REML_RUNTIME_PATH を設定"

(* ========== リンカーコマンド生成 ========== *)

(** プラットフォームに応じたリンカーコマンドを生成
 *
 * @param platform 対象プラットフォーム
 * @param obj_file オブジェクトファイルパス (.o)
 * @param runtime_lib ランタイムライブラリパス (.a)
 * @param output_file 出力実行ファイルパス
 * @return リンカーコマンド文字列
 *)
let generate_linker_command platform obj_file runtime_lib output_file =
  match platform with
  | MacOS ->
      (* macOS: clang を使用 *)
      Printf.sprintf "clang %s %s -o %s -lSystem" obj_file runtime_lib
        output_file
  | Linux ->
      (* Linux: clang または gcc を使用 *)
      Printf.sprintf "clang %s %s -o %s -lc -lm" obj_file runtime_lib
        output_file
  | Windows ->
      (* Windows: lld-link を使用（未実装） *)
      link_error "Windows リンクは Phase 2 で対応予定"
  | Unknown -> link_errorf "サポートされていないプラットフォーム: %s" Sys.os_type

(* ========== 統合リンク関数 ========== *)

(** LLVM IR から実行可能ファイルを生成
 *
 * @param ir_file LLVM IR ファイルパス (.ll)
 * @param output_file 出力実行ファイルパス
 *)
let link_with_runtime ir_file output_file =
  let platform = detect_platform () in
  let runtime_lib = find_runtime_library () in

  (* 1. LLVM IR → オブジェクトファイル *)
  let obj_file = Filename.temp_file "reml_" ".o" in
  let llc_cmd = Printf.sprintf "llc -filetype=obj %s -o %s" ir_file obj_file in
  Printf.eprintf "実行: %s\n%!" llc_cmd;
  let llc_status = Sys.command llc_cmd in
  if llc_status <> 0 then link_errorf "llc でのオブジェクトファイル生成に失敗: %d" llc_status;

  (* 2. オブジェクトファイル + ランタイム → 実行ファイル *)
  let linker_cmd =
    generate_linker_command platform obj_file runtime_lib output_file
  in
  Printf.eprintf "実行: %s\n%!" linker_cmd;
  let link_status = Sys.command linker_cmd in

  (* 一時ファイルを削除 *)
  (try Sys.remove obj_file with _ -> ());

  if link_status <> 0 then link_errorf "リンクに失敗: %d" link_status

(** 簡易リンク（オブジェクトファイルが既にある場合）
 *
 * @param obj_file オブジェクトファイルパス (.o)
 * @param output_file 出力実行ファイルパス
 *)
let link_object_with_runtime obj_file output_file =
  let platform = detect_platform () in
  let runtime_lib = find_runtime_library () in

  let linker_cmd =
    generate_linker_command platform obj_file runtime_lib output_file
  in
  Printf.eprintf "実行: %s\n%!" linker_cmd;
  let link_status = Sys.command linker_cmd in

  if link_status <> 0 then link_errorf "リンクに失敗: %d" link_status

File: compiler/backend/llvm/src/runtime_link.rs
use std::env;
use std::ffi::OsString;
use std::fmt;
use std::fs;
use std::io;
use std::path::{Path, PathBuf};
use std::process::{Command, ExitStatus, Output};
use std::time::{SystemTime, UNIX_EPOCH};

/// 対象プラットフォーム。
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Platform {
    Linux,
    MacOS,
    Windows,
    Unknown,
}

impl Platform {
    /// 環境依存の `os` 名からプラットフォームを判定する。
    pub fn from_os_name(os_name: &str) -> Self {
        match os_name.to_ascii_lowercase().as_str() {
            "linux" => Platform::Linux,
            "macos" | "darwin" => Platform::MacOS,
            "windows" => Platform::Windows,
            _ => Platform::Unknown,
        }
    }

    /// このバイナリを実行している環境のプラットフォーム。
    pub fn detect() -> Self {
        Platform::from_os_name(env::consts::OS)
    }

    /// 人間向けのラベル。
    pub fn label(&self) -> &'static str {
        match self {
            Platform::Linux => "linux",
            Platform::MacOS => "macos",
            Platform::Windows => "windows",
            Platform::Unknown => "unknown",
        }
    }
}

/// リンカー呼び出し情報。
#[derive(Clone, Debug)]
pub struct LinkCommand {
    program: String,
    args: Vec<OsString>,
}

impl LinkCommand {
    /// リンカーで使用する実行ファイル名。
    pub fn program(&self) -> &str {
        &self.program
    }

    /// コマンド引数。
    pub fn args(&self) -> &[OsString] {
        &self.args
    }

    /// `std::process::Command` に変換する。
    fn to_command(&self) -> Command {
        let mut cmd = Command::new(&self.program);
        cmd.args(&self.args);
        cmd
    }
}

/// ランタイムリンク処理で発生するエラー。
#[derive(Debug)]
pub enum RuntimeLinkError {
    /// I/O エラー。
    Io(io::Error),
    /// 指定したランタイムライブラリが見つからない。
    RuntimeLibraryMissing {
        env_value: Option<OsString>,
        candidates: Vec<PathBuf>,
    },
    /// 対応していないプラットフォーム。
    UnsupportedPlatform(String),
    /// 外部コマンドが失敗した。
    CommandFailed {
        program: String,
        exit_status: Option<i32>,
        stdout: String,
        stderr: String,
    },
}

impl fmt::Display for RuntimeLinkError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            RuntimeLinkError::Io(err) => write!(f, "I/O エラー: {}", err),
            RuntimeLinkError::RuntimeLibraryMissing {
                env_value,
                candidates,
            } => {
                write!(
                    f,
                    "ランタイムライブラリが見つかりません (REML_RUNTIME_PATH={:?})。\n\
                     検査した候補: {:?}",
                    env_value, candidates
                )
            }
            RuntimeLinkError::UnsupportedPlatform(platform) => {
                write!(
                    f,
                    "ランタイムリンクはサポートされていないプラットフォーム: {}",
                    platform
                )
            }
            RuntimeLinkError::CommandFailed {
                program,
                exit_status,
                stdout,
                stderr,
            } => write!(
                f,
                "`{}` が失敗 (code={:?}) stdout={} stderr={}",
                program, exit_status, stdout, stderr
            ),
        }
    }
}

impl std::error::Error for RuntimeLinkError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            RuntimeLinkError::Io(err) => Some(err),
            _ => None,
        }
    }
}

impl From<io::Error> for RuntimeLinkError {
    fn from(err: io::Error) -> Self {
        RuntimeLinkError::Io(err)
    }
}

/// ランタイムライブラリの候補パス。
fn runtime_library_candidates() -> Vec<PathBuf> {
    vec![
        PathBuf::from("compiler/runtime/native/build/libreml_runtime.a"),
        PathBuf::from("/usr/local/lib/reml/libreml_runtime.a"),
    ]
}

/// `REML_RUNTIME_PATH` / デフォルト情報に基づいてランタイムライブラリを返す。
pub fn find_runtime_library() -> Result<PathBuf, RuntimeLinkError> {
    let env_value = env::var_os("REML_RUNTIME_PATH");
    if let Some(path) = env_value.as_ref() {
        let candidate = PathBuf::from(path);
        if candidate.exists() {
            return Ok(candidate);
        }
    }
    for candidate in runtime_library_candidates() {
        if candidate.exists() {
            return Ok(candidate);
        }
    }
    Err(RuntimeLinkError::RuntimeLibraryMissing {
        env_value,
        candidates: runtime_library_candidates(),
    })
}

fn describe_execution(output: &Output) -> (String, String) {
    let stdout = String::from_utf8_lossy(&output.stdout).trim().to_string();
    let stderr = String::from_utf8_lossy(&output.stderr).trim().to_string();
    (stdout, stderr)
}

fn map_failure(
    program: &str,
    status: &ExitStatus,
    stdout: String,
    stderr: String,
) -> RuntimeLinkError {
    RuntimeLinkError::CommandFailed {
        program: program.to_string(),
        exit_status: status.code(),
        stdout,
        stderr,
    }
}

fn execute_command(cmd: LinkCommand) -> Result<(), RuntimeLinkError> {
    let output = cmd.to_command().output()?;
    if output.status.success() {
        Ok(())
    } else {
        let (stdout, stderr) = describe_execution(&output);
        Err(map_failure(&cmd.program, &output.status, stdout, stderr))
    }
}

/// LLVM IR ファイルをオブジェクトファイルへ変換する。
pub fn compile_ir_with_llc(ir_file: &Path, obj_file: &Path) -> Result<(), RuntimeLinkError> {
    let output = Command::new("llc")
        .arg("-filetype=obj")
        .arg(ir_file)
        .arg("-o")
        .arg(obj_file)
        .output()?;
    if output.status.success() {
        Ok(())
    } else {
        let (stdout, stderr) = describe_execution(&output);
        Err(map_failure("llc", &output.status, stdout, stderr))
    }
}

/// オブジェクトファイルとランタイムをリンクするコマンドを生成する。
pub fn generate_link_command(
    platform: Platform,
    obj_file: &Path,
    runtime_lib: &Path,
    output_file: &Path,
) -> Result<LinkCommand, RuntimeLinkError> {
    let mut args = Vec::new();
    match platform {
        Platform::MacOS => {
            args.push(obj_file.as_os_str().to_owned());
            args.push(runtime_lib.as_os_str().to_owned());
            args.push(OsString::from("-o"));
            args.push(output_file.as_os_str().to_owned());
            args.push(OsString::from("-lSystem"));
            Ok(LinkCommand {
                program: "clang".into(),
                args,
            })
        }
        Platform::Linux => {
            args.push(obj_file.as_os_str().to_owned());
            args.push(runtime_lib.as_os_str().to_owned());
            args.push(OsString::from("-o"));
            args.push(output_file.as_os_str().to_owned());
            args.push(OsString::from("-lc"));
            args.push(OsString::from("-lm"));
            Ok(LinkCommand {
                program: "clang".into(),
                args,
            })
        }
        Platform::Windows => Err(RuntimeLinkError::UnsupportedPlatform("windows".into())),
        Platform::Unknown => Err(RuntimeLinkError::UnsupportedPlatform("unknown".into())),
    }
}

/// オブジェクトファイルとランタイムライブラリをリンクする。
pub fn link_object_with_runtime(
    obj_file: &Path,
    output_file: &Path,
    platform: Platform,
    runtime_lib: &Path,
) -> Result<(), RuntimeLinkError> {
    let link_cmd = generate_link_command(platform, obj_file, runtime_lib, output_file)?;
    execute_command(link_cmd)
}

/// LLVM IR から実行ファイルを生成し、リンク後のオブジェクトファイルを削除する。
pub fn link_with_runtime(ir_file: &Path, output_file: &Path) -> Result<(), RuntimeLinkError> {
    let obj_file = env::temp_dir().join(link_object_name());
    let runtime_lib = find_runtime_library()?;
    let platform = Platform::detect();
    let result = compile_ir_with_llc(ir_file, &obj_file)
        .and_then(|_| link_object_with_runtime(&obj_file, output_file, platform, &runtime_lib));
    let _ = fs::remove_file(&obj_file);
    result
}

fn link_object_name() -> String {
    let timestamp = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|value| value.as_nanos())
        .unwrap_or_default();
    format!("reml_runtime_{}.o", timestamp)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::env;
    use std::fs;

    #[test]
    fn platform_from_name() {
        assert_eq!(Platform::from_os_name("linux"), Platform::Linux);
        assert_eq!(Platform::from_os_name("macos"), Platform::MacOS);
        assert_eq!(Platform::from_os_name("darwin"), Platform::MacOS);
        assert_eq!(Platform::from_os_name("windows"), Platform::Windows);
        assert_eq!(Platform::from_os_name("unknown-os"), Platform::Unknown);
    }

    #[test]
    fn generate_link_command_linux() {
        let platform = Platform::Linux;
        let obj = PathBuf::from("/tmp/test.o");
        let runtime = PathBuf::from("/tmp/libreml_runtime.a");
        let output = PathBuf::from("/tmp/result");
        let cmd = generate_link_command(platform, &obj, &runtime, &output).unwrap();
        assert_eq!(cmd.program(), "clang");
        let args: Vec<String> = cmd
            .args()
            .iter()
            .map(|arg| arg.to_string_lossy().to_string())
            .collect();
        assert!(args.contains(&obj.to_string_lossy().to_string()));
        assert!(args.contains(&runtime.to_string_lossy().to_string()));
        assert!(args.contains(&output.to_string_lossy().to_string()));
        assert!(args.contains(&"-lc".to_string()));
        assert!(args.contains(&"-lm".to_string()));
    }

    #[test]
    fn generate_link_command_macos() {
        let platform = Platform::MacOS;
        let obj = PathBuf::from("/tmp/test.o");
        let runtime = PathBuf::from("/tmp/libreml_runtime.a");
        let output = PathBuf::from("/tmp/result");
        let cmd = generate_link_command(platform, &obj, &runtime, &output).unwrap();
        assert_eq!(cmd.program(), "clang");
        let args: Vec<String> = cmd
            .args()
            .iter()
            .map(|arg| arg.to_string_lossy().to_string())
            .collect();
        assert!(args.contains(&"-lSystem".to_string()));
    }

    #[test]
    fn find_runtime_library_prefers_env() {
        let temp_path = env::temp_dir().join("reml_runtime_test_dummy.a");
        fs::write(&temp_path, b"dummy").unwrap();
        env::set_var("REML_RUNTIME_PATH", &temp_path);
        let found = find_runtime_library().expect("Runtime path should be found");
        assert_eq!(found, temp_path);
        env::remove_var("REML_RUNTIME_PATH");
        let _ = fs::remove_file(&temp_path);
    }
}
